### PR TITLE
Support github as a script storage

### PR DESCRIPTION
**Pull request by donggyu04**
Support git for script storage, user can select script from `GIT` or embedded `SVN` both.

### CODE REVIEW

File: ngrinder-controller/src/main/java/org/ngrinder/script/model/FileType.java [hunk 1]
```diff
@@ -39,6 +39,8 @@
 	CSV("csv", "csv", FileCategory.DATA, false, true),
 	/** JSON. */
 	JSON("json", "json", FileCategory.DATA, false, true),
+	/** YAML for github config. */
+	YAML("yaml", "yml", FileCategory.DATA, false, false),
```
Comment (imbyungjun) [inline ngrinder-controller/src/main/java/org/ngrinder/script/model/FileType.java]: What about just "YAML"? YAML is not limited to github configuration.

File: ngrinder-controller/src/main/java/org/ngrinder/common/constant/CacheConstants.java [hunk 2]
```diff
@@ -40,5 +40,7 @@
 
 	String CACHE_USER_ENTITY = "org.ngrinder.model.User";
 
+	String CACHE_GIT_SCRIPTS = "git_scripts";
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/common/constant/CacheConstants.java]: I think it's better to name it github not git.

File: ngrinder-controller/src/main/java/org/ngrinder/infra/config/WebMvcConfig.java [hunk 3]
```diff
@@ -142,4 +143,9 @@ public Jackson2ObjectMapperBuilderCustomizer jackson2ObjectMapperBuilderCustomiz
 			customizer.modules(new NumberModule());
 		};
 	}
+
+	@Bean
+	public RestTemplate restTemplate() {
+		return new RestTemplate();
+	}
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/infra/config/WebMvcConfig.java]: Why do we need this?

File: ngrinder-controller/src/main/java/org/ngrinder/script/model/GitApi.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+	GitApi(String apiTemplate) {
+		this.apiTemplate = apiTemplate;
+	}
+
+	public String buildUrl(GitConfig gitConfig, Object... args) {
+		String[] baseArgs = { gitConfig.getBaseUrl(), gitConfig.getOwner(), gitConfig.getRepo() };
+		return format(apiTemplate, addAll(baseArgs, args));
+	}
+}
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/script/model/GitApi.java]: Why don't you use github api library like this.
https://github-api.kohsuke.org/

File: ngrinder-controller/src/main/java/org/ngrinder/script/service/FileEntryService.java [hunk 5]
```diff
@@ -127,12 +129,29 @@ public void prepare(User user) {
 		try {
 			if (!newUserDirectory.exists()) {
 				createUserRepo(user, newUserDirectory);
+				createGitConfig(user);
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/script/service/FileEntryService.java]: name it github config

File: ngrinder-controller/src/main/java/org/ngrinder/script/service/FileEntryService.java [hunk 6]
```diff
@@ -127,12 +129,29 @@ public void prepare(User user) {
...
 			}
 		} catch (SVNException e) {
 			LOG.error("Error while prepare user {}'s repo", user.getUserName(), e);
 		}
 	}
 
+	private void createGitConfig(User user) {
+		try {
+			String content = IOUtils.toString(gitConfigTemplate.getURL());
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/script/service/FileEntryService.java]: Please follow the standard way to copy file. see Config

File: ngrinder-controller/src/main/java/org/ngrinder/script/service/GitService.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	private Map getAllFileAndFolder(GitConfig gitConfig, String sha) {
+		return execute(TREES.buildUrl(gitConfig, sha)).getBody();
+	}
+
+	private ResponseEntity<Map> execute(String apiUrl) {
+		return restTemplate.exchange(apiUrl, GET, httpEntity, Map.class);
+	}
+}
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/script/service/GitService.java]: Why about integrate this into FileEntryService?

File: ngrinder-controller/src/main/java/org/ngrinder/script/service/FileEntryService.java [hunk 8]
```diff
@@ -85,24 +88,23 @@
 
 	private Cache fileEntryCache;
 
+	@Value("classpath:gitconfig_template/gitconfig.yml")
+	private Resource gitConfigTemplate;
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/script/service/FileEntryService.java]: Please do not copy it from classpath. instead copy it from somewhere in .ngrinder folder

File: ngrinder-controller/src/main/java/org/ngrinder/perftest/controller/PerfTestApiController.java [hunk 9]
```diff
@@ -13,6 +13,7 @@
  */
 package org.ngrinder.perftest.controller;
 
+import lombok.RequiredArgsConstructor;
```
Comment (junoyoon) [inline ngrinder-controller/src/main/java/org/ngrinder/perftest/controller/PerfTestApiController.java]: Whenever u change the import order is changed.

@imbyungjun  @donggyu04 do you guys same style?

Comment (donggyu04) [reply ngrinder-controller/src/main/java/org/ngrinder/perftest/controller/PerfTestApiController.java]: We'll check IDE settings.

File: ngrinder-controller/src/main/java/org/ngrinder/script/model/GitApi.java (hunk 4 above)
Comment (donggyu04) [reply ngrinder-controller/src/main/java/org/ngrinder/script/model/GitApi.java]: I modified to use this.

File: ngrinder-controller/src/main/java/org/ngrinder/infra/config/Config.java [hunk 10]
```diff
@@ -601,6 +604,26 @@ public String getProcessAndThreadPolicyScript() {
 		return policyScript;
 	}
 
+
+	private String gitHubConfigTemplate = "";
```
[the remark this answers is not in the document]
Comment (donggyu04) [reply ngrinder-controller/src/main/java/org/ngrinder/infra/config/Config.java]: This cache is used every time to creating `.gitconfig.yml`.

**Comment by donggyu04** [comment]
Can one of the admins verify this patch?

File: ngrinder-frontend/src/js/components/perftest/detail/Config.vue [hunk 11]
```diff
@@ -317,7 +319,7 @@
             if (this.test.config.scm === 'addGitHub') {
                 this.createGitConfig();
             } else {
-                this.display.showGitHubRefreshBtn = this.test.config.scm !== 'svn';
+                this.display.showGitHubRefreshBtn = this.isGitHubStorage();
```
Comment (imbyungjun) [inline ngrinder-frontend/src/js/components/perftest/detail/Config.vue]: `display.show...` looks redundant. How about just `display.githubRefeshBtn`?

File: ngrinder-frontend/src/js/components/perftest/detail/Config.vue [hunk 12]
```diff
@@ -428,21 +431,28 @@
...
+                return;
+            }
+
+            if (this.$refs.scriptSelect.getSelectedOptionValidate() !== '-1') {
+                this.test.config.scriptRevision = revision;
+                this.getTargetHosts();
+                this.getScriptResource();
+                this.display.showRevisionBtn = true;
+            } else {
+                this.test.config.scriptRevision = -1;
```
Comment (imbyungjun) [inline ngrinder-frontend/src/js/components/perftest/detail/Config.vue]: scriptRevision should be '-1' now?

File: ngrinder-controller/src/main/java/org/ngrinder/perftest/service/PerfTestService.java [hunk 13]
```diff
@@ -628,14 +634,30 @@ public GrinderProperties getGrinderProperties(PerfTest perfTest, ScriptHandler s
...
+		if (isGitHubScript(scm)) {
+			String gitHubConfigName = extractGitHubConfigName(scm);
+			String scriptName = perfTest.getScriptName();
+
+			GitHubConfig gitHubConfig = gitHubFileEntryService.getGitHubConfig(user, gitHubConfigName);
+			GitHub gitHub = gitHubFileEntryService.getGitHubClient(gitHubConfig);
+			GHRepository ghRepository = gitHub.getRepository(gitHubConfig.getOwner() + "/" + gitHubConfig.getRepo());
+
+			gitHubFileEntryService.checkoutGitHubScript(user, ghRepository, gitHubConfig, scriptName);
+			scriptEntry = gitHubFileEntryService.getOne(user, ghRepository, gitHubConfigName, scriptName);
```
Comment (imbyungjun) [inline ngrinder-controller/src/main/java/org/ngrinder/perftest/service/PerfTestService.java]: I dont' think that PerftestService have to know about either `GithubClient` or `GHRepository`. Is it possible to change `githubFileEntryService.getOne(user, githubConfig, scriptName)`?

Comment (donggyu04) [reply ngrinder-controller/src/main/java/org/ngrinder/perftest/service/PerfTestService.java]: To get `GHRepository` it needs to github api call.
so. i reuse this to reduce api call.

File: ngrinder-frontend/src/js/components/perftest/detail/Config.vue (hunk 12 above)
Comment (donggyu04) [reply ngrinder-frontend/src/js/components/perftest/detail/Config.vue]: When change script, `svn` script always set to `HEAD` revision.

File: ngrinder-frontend/src/js/components/perftest/detail/Config.vue [hunk 14]
```diff
@@ -743,6 +742,20 @@
             return pathToken[pathToken.length - 1];
         }
 
+        getShortPath(path) {
+            if (!path) {
+                return '';
+            }
+
+            if (path.length >= 40 && path.includes('/')) {
```
Comment (songeunwoo) [inline ngrinder-frontend/src/js/components/perftest/detail/Config.vue]: How about improving with css?

File: ngrinder-controller/src/main/java/org/ngrinder/script/model/GitHubConfig.java [hunk 15]
```diff
@@ -65,5 +67,34 @@ public GitHubConfig deserialize(JsonParser jsonParser, DeserializationContext ct
...
+		 * Check gitHub configuration has unsupported fields.
+		 *
+		 * @throws InvalidGitHubConfigurationException occurs when unsupported field is exist.
+		 * */
+		private void checkUnsupportedField(JsonNode jsonNode) {
+			Iterator<Entry<String, JsonNode>> fields = jsonNode.fields();
+			while (fields.hasNext()) {
+				Entry<String, JsonNode> field = fields.next();
+				String key = field.getKey();
+				if (!isSupportedField(key)) {
```
Comment (donggyu04) [inline ngrinder-controller/src/main/java/org/ngrinder/script/model/GitHubConfig.java:93]: Notify error message  if user set unsupported field in github configuration file

image

Comment (songeunwoo) [reply ngrinder-controller/src/main/java/org/ngrinder/script/model/GitHubConfig.java:93]: good.

**Pull request merged**